File: DotNet/CesiumLanguageWriterTests/TestNodeTransformationsCesiumWriter.cs
using System;
using System.Collections.Generic;
using System.IO;
using CesiumLanguageWriter;
using CesiumLanguageWriter.Advanced;
using NUnit.Framework;

namespace CesiumLanguageWriterTests
{
    [TestFixture]
    public class TestNodeTransformationsCesiumWriter : TestCesiumPropertyWriter<NodeTransformationsCesiumWriter>
    {
        [Test]
        public void SingleNodeTransformation()
        {
            const string expectedNodeName = "node1";
            var expectedScale = Cartesian.Zero;
            var expectedRotation = UnitQuaternion.Identity;
            var expectedTranslation = new Cartesian(3.0, 3.0, 3.0);

            using (var packet = OpenPacket())
            using (var model = packet.OpenModelProperty())
            using (var nodeTransformations = model.OpenNodeTransformationsProperty())
            {
                using (var nodeTransformation = nodeTransformations.OpenNodeTransformationProperty(expectedNodeName))
                {
                    nodeTransformation.WriteScaleProperty(expectedScale);
                    nodeTransformation.WriteRotationProperty(expectedRotation);
                    nodeTransformation.WriteTranslationProperty(expectedTranslation);
                }
            }

            AssertExpectedJson(PacketCesiumWriter.ModelPropertyName, new Dictionary<string, object>
            {
                {
                    ModelCesiumWriter.NodeTransformationsPropertyName, new Dictionary<string, object>
                    {
                        {
                            expectedNodeName, new Dictionary<string, object>
                            {
                                { NodeTransformationCesiumWriter.ScalePropertyName, expectedScale },
                                { NodeTransformationCesiumWriter.RotationPropertyName, expectedRotation },
                                { NodeTransformationCesiumWriter.TranslationPropertyName, expectedTranslation },
                            }
                        },
                    }
                },
            });
        }

        [Test]
        public void MultipleNodeTransformations()
        {
            const string expectedNodeName1 = "node1";
            var expectedScale1 = new Cartesian(1.0, 2.0, 3.0);
            var expectedRotation1 = new UnitQuaternion(1.0, 0.0, 0.0, 0.0);
            var expectedTranslation1 = new Cartesian(4.0, 5.0, 6.0);
            const string expectedNodeName2 = "node2";
            var expectedScale2 = new Cartesian(4.0, 5.0, 6.0);
            var expectedRotation2 = new UnitQuaternion(0.0, 0.0, 0.0, 1.0);
            var expectedTranslation2 = new Cartesian(7.0, 8.0, 9.0);

            using (var packet = OpenPacket())
            using (var model = packet.OpenModelProperty())
            using (var nodeTransformations = model.OpenNodeTransformationsProperty())
            {
                using (var nodeTransformation = nodeTransformations.OpenNodeTransformationProperty(expectedNodeName1))
                {
                    nodeTransformation.WriteScaleProperty(expectedScale1);
                    nodeTransformation.WriteRotationProperty(expectedRotation1);
                    nodeTransformation.WriteTranslationProperty(expectedTranslation1);
                }

                using (var nodeTransformation = nodeTransformations.OpenNodeTransformationProperty(expectedNodeName2))
                {
                    nodeTransformation.WriteScaleProperty(expectedScale2);
                    nodeTransformation.WriteRotationProperty(expectedRotation2);
                    nodeTransformation.WriteTranslationProperty(expectedTranslation2);
                }
            }

            AssertExpectedJson(PacketCesiumWriter.ModelPropertyName, new Dictionary<string, object>
            {
                {
                    ModelCesiumWriter.NodeTransformationsPropertyName, new Dictionary<string, object>
                    {
                        {
                            expectedNodeName1, new Dictionary<string, object>
                            {
                                { NodeTransformationCesiumWriter.ScalePropertyName, expectedScale1 },
                                { NodeTransformationCesiumWriter.RotationPropertyName, expectedRotation1 },
                                { NodeTransformationCesiumWriter.TranslationPropertyName, expectedTranslation1 },
                            }
                        },
                        {
                            expectedNodeName2, new Dictionary<string, object>
                            {
                                { NodeTransformationCesiumWriter.ScalePropertyName, expectedScale2 },
                                { NodeTransformationCesiumWriter.RotationPropertyName, expectedRotation2 },
                                { NodeTransformationCesiumWriter.TranslationPropertyName, expectedTranslation2 },
                            }
                        },
                    }
                },
            });
        }

        [Test]
        public void CreateExampleFile()
        {
            using (var stringWriter = new StringWriter())
            {
                var output = new CesiumOutputStream(stringWriter, true);
                var writer = new CesiumStreamWriter();
                using (var packet = writer.OpenPacket(output))
                {
                    packet.WriteId("MyID");

                    var startDate = new GregorianDate(2012, 4, 2, 12, 0, 0).ToJulianDate();
                    var stopDate = new GregorianDate(2012, 4, 2, 12, 1, 0).ToJulianDate();

                    using (var clock = packet.OpenClockProperty())
                    {
                        using (var interval = clock.OpenInterval(startDate, stopDate))
                        {
                            interval.WriteCurrentTime(startDate);
                        }
                    }

                    using (var model = packet.OpenModelProperty())
                    {
                        model.WriteGltfProperty(new Uri("example.gltf", UriKind.Relative), CesiumResourceBehavior.LinkTo);

                        using (var nodeTransformations = model.OpenNodeTransformationsProperty())
                        {
                            using (var nodeTransformation = nodeTransformations.OpenNodeTransformationProperty("node1"))
                            {
                                nodeTransformation.WriteScaleProperty(new Cartesian(1.0, 2.0, 3.0));
                                nodeTransformation.WriteRotationProperty(UnitQuaternion.Identity);
                                nodeTransformation.WriteTranslationProperty(new Cartesian(4.0, 5.0, 6.0));
                            }

                            using (var nodeTransformation = nodeTransformations.OpenNodeTransformationProperty("node2"))
                            {
                                var dates = new List<JulianDate>
                                {
                                    startDate,
                                    stopDate
                                };
                                var values = new List<Cartesian>
                                {
                                    new Cartesian(1.0, 2.0, 3.0),
                                    new Cartesian(10.0, 12.0, 14.0)
                                };
                                nodeTransformation.WriteScaleProperty(dates, values);
                            }
                        }
                    }
                }

                Console.WriteLine(stringWriter.ToString());
            }
        }

        protected override CesiumPropertyWriter<NodeTransformationsCesiumWriter> CreatePropertyWriter(string propertyName)
        {
            return new NodeTransformationsCesiumWriter(propertyName);
        }
    }
}

File: Java/CesiumLanguageWriterTests/translatedSrc/cesiumlanguagewritertests/TestNodeTransformationsCesiumWriter.java
package cesiumlanguagewritertests;


import agi.foundation.compatibility.*;
import agi.foundation.compatibility.ConsoleHelper;
import agi.foundation.compatibility.MapHelper;
import agi.foundation.compatibility.TestContextRule;
import agi.foundation.compatibility.UriHelper;
import agi.foundation.compatibility.UriKind;
import agi.foundation.compatibility.Using;
import cesiumlanguagewriter.*;
import cesiumlanguagewriter.advanced.*;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Map;
import javax.annotation.Nonnull;
import org.junit.ClassRule;
import org.junit.FixMethodOrder;
import org.junit.Rule;
import org.junit.runners.MethodSorters;
import org.junit.Test;

@SuppressWarnings({
    "unused",
    "deprecation",
    "serial"
})
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class TestNodeTransformationsCesiumWriter extends TestCesiumPropertyWriter<NodeTransformationsCesiumWriter> {
    @Test
    public final void singleNodeTransformation() {
        final String expectedNodeName = "node1";
        Cartesian expectedScale = Cartesian.getZero();
        UnitQuaternion expectedRotation = UnitQuaternion.getIdentity();
        Cartesian expectedTranslation = new Cartesian(3.0, 3.0, 3.0);
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<ModelCesiumWriter> using$1 = new Using<ModelCesiumWriter>(packet.openModelProperty())) {
                final ModelCesiumWriter model = using$1.resource;
                try (Using<NodeTransformationsCesiumWriter> using$2 = new Using<NodeTransformationsCesiumWriter>(model.openNodeTransformationsProperty())) {
                    final NodeTransformationsCesiumWriter nodeTransformations = using$2.resource;
                    try (Using<NodeTransformationCesiumWriter> using$3 = new Using<NodeTransformationCesiumWriter>(nodeTransformations.openNodeTransformationProperty(expectedNodeName))) {
                        final NodeTransformationCesiumWriter nodeTransformation = using$3.resource;
                        nodeTransformation.writeScaleProperty(expectedScale);
                        nodeTransformation.writeRotationProperty(expectedRotation);
                        nodeTransformation.writeTranslationProperty(expectedTranslation);
                    }
                }
            }
        }
        final Map<String, Object> tempCollection$2 = MapHelper.create();
        MapHelper.add(tempCollection$2, NodeTransformationCesiumWriter.ScalePropertyName, expectedScale);
        MapHelper.add(tempCollection$2, NodeTransformationCesiumWriter.RotationPropertyName, expectedRotation);
        MapHelper.add(tempCollection$2, NodeTransformationCesiumWriter.TranslationPropertyName, expectedTranslation);
        final Map<String, Object> tempCollection$1 = MapHelper.create();
        MapHelper.add(tempCollection$1, expectedNodeName, tempCollection$2);
        final Map<String, Object> tempCollection$0 = MapHelper.create();
        MapHelper.add(tempCollection$0, ModelCesiumWriter.NodeTransformationsPropertyName, tempCollection$1);
        assertExpectedJson(PacketCesiumWriter.ModelPropertyName, tempCollection$0);
    }

    @Test
    public final void multipleNodeTransformations() {
        final String expectedNodeName1 = "node1";
        Cartesian expectedScale1 = new Cartesian(1.0, 2.0, 3.0);
        UnitQuaternion expectedRotation1 = new UnitQuaternion(1.0, 0.0, 0.0, 0.0);
        Cartesian expectedTranslation1 = new Cartesian(4.0, 5.0, 6.0);
        final String expectedNodeName2 = "node2";
        Cartesian expectedScale2 = new Cartesian(4.0, 5.0, 6.0);
        UnitQuaternion expectedRotation2 = new UnitQuaternion(0.0, 0.0, 0.0, 1.0);
        Cartesian expectedTranslation2 = new Cartesian(7.0, 8.0, 9.0);
        try (Using<PacketCesiumWriter> using$0 = new Using<PacketCesiumWriter>(openPacket())) {
            final PacketCesiumWriter packet = using$0.resource;
            try (Using<ModelCesiumWriter> using$1 = new Using<ModelCesiumWriter>(packet.openModelProperty())) {
                final ModelCesiumWriter model = using$1.resource;
                try (Using<NodeTransformationsCesiumWriter> using$2 = new Using<NodeTransformationsCesiumWriter>(model.openNodeTransformationsProperty())) {
                    final NodeTransformationsCesiumWriter nodeTransformations = using$2.resource;
                    try (Using<NodeTransformationCesiumWriter> using$3 = new Using<NodeTransformationCesiumWriter>(nodeTransformations.openNodeTransformationProperty(expectedNodeName1))) {
                        final NodeTransformationCesiumWriter nodeTransformation = using$3.resource;
                        nodeTransformation.writeScaleProperty(expectedScale1);
                        nodeTransformation.writeRotationProperty(expectedRotation1);
                        nodeTransformation.writeTranslationProperty(expectedTranslation1);
                    }
                    try (Using<NodeTransformationCesiumWriter> using$4 = new Using<NodeTransformationCesiumWriter>(nodeTransformations.openNodeTransformationProperty(expectedNodeName2))) {
                        final NodeTransformationCesiumWriter nodeTransformation = using$4.resource;
                        nodeTransformation.writeScaleProperty(expectedScale2);
                        nodeTransformation.writeRotationProperty(expectedRotation2);
                        nodeTransformation.writeTranslationProperty(expectedTranslation2);
                    }
                }
            }
        }
        final Map<String, Object> tempCollection$2 = MapHelper.create();
        MapHelper.add(tempCollection$2, NodeTransformationCesiumWriter.ScalePropertyName, expectedScale1);
        MapHelper.add(tempCollection$2, NodeTransformationCesiumWriter.RotationPropertyName, expectedRotation1);
        MapHelper.add(tempCollection$2, NodeTransformationCesiumWriter.TranslationPropertyName, expectedTranslation1);
        final Map<String, Object> tempCollection$3 = MapHelper.create();
        MapHelper.add(tempCollection$3, NodeTransformationCesiumWriter.ScalePropertyName, expectedScale2);
        MapHelper.add(tempCollection$3, NodeTransformationCesiumWriter.RotationPropertyName, expectedRotation2);
        MapHelper.add(tempCollection$3, NodeTransformationCesiumWriter.TranslationPropertyName, expectedTranslation2);
        final Map<String, Object> tempCollection$1 = MapHelper.create();
        MapHelper.add(tempCollection$1, expectedNodeName1, tempCollection$2);
        MapHelper.add(tempCollection$1, expectedNodeName2, tempCollection$3);
        final Map<String, Object> tempCollection$0 = MapHelper.create();
        MapHelper.add(tempCollection$0, ModelCesiumWriter.NodeTransformationsPropertyName, tempCollection$1);
        assertExpectedJson(PacketCesiumWriter.ModelPropertyName, tempCollection$0);
    }

    @Test
    public final void createExampleFile() {
        try (Using<StringWriter> using$0 = new Using<StringWriter>(new StringWriter())) {
            final StringWriter stringWriter = using$0.resource;
            CesiumOutputStream output = new CesiumOutputStream(stringWriter, true);
            CesiumStreamWriter writer = new CesiumStreamWriter();
            try (Using<PacketCesiumWriter> using$1 = new Using<PacketCesiumWriter>(writer.openPacket(output))) {
                final PacketCesiumWriter packet = using$1.resource;
                packet.writeId("MyID");
                JulianDate startDate = new GregorianDate(2012, 4, 2, 12, 0, 0.0).toJulianDate();
                JulianDate stopDate = new GregorianDate(2012, 4, 2, 12, 1, 0.0).toJulianDate();
                try (Using<ClockCesiumWriter> using$2 = new Using<ClockCesiumWriter>(packet.openClockProperty())) {
                    final ClockCesiumWriter clock = using$2.resource;
                    try (Using<ClockCesiumWriter> using$3 = new Using<ClockCesiumWriter>(clock.openInterval(startDate, stopDate))) {
                        final ClockCesiumWriter interval = using$3.resource;
                        interval.writeCurrentTime(startDate);
                    }
                }
                try (Using<ModelCesiumWriter> using$4 = new Using<ModelCesiumWriter>(packet.openModelProperty())) {
                    final ModelCesiumWriter model = using$4.resource;
                    model.writeGltfProperty(UriHelper.create("example.gltf", UriKind.RELATIVE), CesiumResourceBehavior.LINK_TO);
                    try (Using<NodeTransformationsCesiumWriter> using$5 = new Using<NodeTransformationsCesiumWriter>(model.openNodeTransformationsProperty())) {
                        final NodeTransformationsCesiumWriter nodeTransformations = using$5.resource;
                        try (Using<NodeTransformationCesiumWriter> using$6 = new Using<NodeTransformationCesiumWriter>(nodeTransformations.openNodeTransformationProperty("node1"))) {
                            final NodeTransformationCesiumWriter nodeTransformation = using$6.resource;
                            nodeTransformation.writeScaleProperty(new Cartesian(1.0, 2.0, 3.0));
                            nodeTransformation.writeRotationProperty(UnitQuaternion.getIdentity());
                            nodeTransformation.writeTranslationProperty(new Cartesian(4.0, 5.0, 6.0));
                        }
                        try (Using<NodeTransformationCesiumWriter> using$7 = new Using<NodeTransformationCesiumWriter>(nodeTransformations.openNodeTransformationProperty("node2"))) {
                            final NodeTransformationCesiumWriter nodeTransformation = using$7.resource;
                            final ArrayList<JulianDate> tempCollection$0 = new ArrayList<JulianDate>();
                            tempCollection$0.add(startDate);
                            tempCollection$0.add(stopDate);
                            ArrayList<JulianDate> dates = tempCollection$0;
                            final ArrayList<Cartesian> tempCollection$1 = new ArrayList<Cartesian>();
                            tempCollection$1.add(new Cartesian(1.0, 2.0, 3.0));
                            tempCollection$1.add(new Cartesian(10.0, 12.0, 14.0));
                            ArrayList<Cartesian> values = tempCollection$1;
                            nodeTransformation.writeScaleProperty(dates, values);
                        }
                    }
                }
            }
            ConsoleHelper.writeLine(stringWriter.toString());
        }
    }

    @Override
    protected CesiumPropertyWriter<NodeTransformationsCesiumWriter> createPropertyWriter(String propertyName) {
        return new NodeTransformationsCesiumWriter(propertyName);
    }

    @Nonnull
    private static final TestContextRule rule$testContext = new TestContextRule();

    @Nonnull
    @Rule
    @ClassRule
    public static TestContextRule getRule$testContext() {
        return rule$testContext;
    }
}